# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1303 Депрейс Александр. Вариант №7.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения
### Аддитивная цепочка
1. Аддитивной цепочкой для $n \in \mathbb{N}$ называется последовательность натуральных чисел, где каждый элемент последовательности равен сумме каких либо двух предыдущих:<br>
$$1 = a_0,a_1,a_2,\dots,a_r = n \\
a_i = a_j + a_k, \quad k \le j \lt i, \quad i = 1,2,\dots,r$$


2. $l(n) = r$ - наименьшая длина $r$, для которой существует аддитивная цепочка для $n$.


3. Пара $(j,k), \quad 0 \le k \le j \lt i$ называется шагом $i$


4. Если существует более чем одна пара $(j, k)$, полагаем, что $j$ - наибольшая из всех возможных. 


5. Виды шагов:<br>
    1) Удвоение:$j = k = i - 1$<br>
    2) Звездный шаг: $j = i - 1$<br>
    3) Малый шаг: $\lambda(a_i) = \lambda(a_{i-1})$
    
    
6. Свойства:<br>
    1) Шаг 1 - всегда удвоение<br>
    2) Удвоение - звездный шаг, но никогда не малый шаг<br>
    3) За удвоением всегда следует звездный шаг.<br>
    4) Если $i$ый шаг не малый, то $(i + 1)$ый шаг либо малый, либо звездный, либо и тот и другой.<br>
    5) Если $(i + 1)$ый шаг не звездный и не малый, то $i$ый шаг должен быть малым шагом.
    
    
7. Теорема: Если аддитивная цепочка включает $d$ удвоений и $f = r - d$ неудвоений, то<br>
    $$n \le 2^{d - 1}\cdot F_{f + 3}$$, где $F_{j}$ - число Фиббоначи на позиции $j$.
    

8. Следствие: Если аддитивная цепочка включает $а$ неудвоений и $s$ малых шагов, то<br>
    $$s \le f \le \frac{s}{(1-\log_2(\varphi))}$$, где $\varphi = \frac{\sqrt{5} + 1}{2}$ - золотое сечение
    
    
9. Алгоритм Брауэра позволяет вычислить $n$-ую степень за<br>
    $$l_{B}(n) = \lambda(n) + \frac{\lambda(n)}{\lambda\lambda(n)} + O\left(\frac{\lambda(n)\lambda\lambda\lambda(n)}{(\lambda\lambda(n))^{2}}\right)$$ умноженией.


10. Самая короткая аддитивная цепочка для числа $n$ имеет длину не более $\lambda(n)$.


11. Доказано, что почти для всех $n$ минимальная аддитивная цепочка имеет длину $l_{B}(n)$.


### Алгоритм Брауэра
Для $n \in \mathbb{N}$ при заданном $k \in \mathbb{N}$ можно построить цепочку Брауэра с помощью рекуррентной формулы:<br>
$$ B_{k}(n) =
  \begin{cases}
    1,2,3,\dots,2^{k} - 1       & \quad n < 2^{k}\\
    B_{k}(q),2q,4q,8q,\dots,2^{k}q,n  & \quad n \ge 2^{k}, q = \lfloor\frac{n}{2^{k}}\rfloor \\
  \end{cases}
$$
Данная цепочка будет иметь длину:
$$ l_{B}(n) = j(k + 1) + 2^{k} - 2, $$ при условии, что $jk \le \lambda(n) \lt (j + 1)k$

Длина будет минимизирована  для больших $n$, если положить, что $k = \lambda\lambda(n) - 2\lambda\lambda\lambda(n)$

Сам алгоритм:

1. Задается некий фиксируемый $k$ для рассматривемого $n$. Выполняется вычисление "вспомогательных чисел":
$$d = 2^{k}, q_1 = \left[\frac{n}{d}\right], \quad r_1 = n \, \mathrm{mod} \, d \Rightarrow n = q_1 d + r_1 \quad(0 \le r_1 \lt d)$$
$$q_2 = \left[\frac{q_1}{d}\right], \quad r_2 = q_1 \, \mathrm{mod} \, d \Rightarrow q_1 = q_2 d + r_2 \quad(0 \le r_2 \lt d)$$

2. Данная процедура продолжается до тех пор, пока не появится такое $q_s \lt d$. Следовательно $q_{s-1} = q_s d + r_s$

3. Таким образом, n имеет вид:
$$n = 2^{k}q_1 + r_1 = 2^{k}\left(2^{k}q_2 + r_2\right) + r_1$ = \dots = \\ 2^{k}\left(2^{k} \left(\dots\left(2^{k}q_s + r_s\right)\dots\right) + r_2\right) + r_1 $$

### Звездная цепочка
Звездная цепочка - аддитивная цепочка, которая содержит только звездные шаги.

1. Длина звездной цепочки: $l^*(n)$
2. Шаг звездной цепочки: $a_i = a_{i - 1} + a_k \quad \forall k < i$
3. Очевидно, что: $l(n) \le l^*(n)$

Пусть задана звездная цепочка длины $m - 1$ вида $1 = a_1,a_2,\dots,a_m$.<br>
Для каждой звездной цепочки существует вектор индексов $r = \left\{ r_1,r_2, \dots ,r_m \right\}$ длины $m - 1$ такой, что $r_i = \left\{ z: \: 1 \le z \le i \right\} \quad a_i = a_{i - 1} + a_{r_{i - 1}}, \; a \le i \le m - 1$
1. Первый элемент вектора $r_1$ всегда $= 1$, второй элемент $r_2 = \left\{1,2\right\}$. Последний элемент принимает значения от $1$ до $m-1$
2. Наибольшая звездная цепочка будет иметь вид $a = 1,2,4,\dots,2^{m - 1}$, а ее вектор индексов $r = \left\{1,2,3,\dots,m-1\right\}$
3. Наименьшая звездная цепочка будет иметь вид $a = 1,2,3,\dots,m$, а ее вектор индексов $r = \left\{1,1,1,\dots,1\right\}$

### Сравнимость двух векторов
Даны 2 вектора индексов равной длины $r = \left\{r_1,r_2,r_3,\dots,r_m\right\}, \widetilde{r} = \left\{\widetilde{r_1},\widetilde{r_2},\widetilde{r_3},\dots,\widetilde{r_m}\right\}$. Тогда $r \succ \widetilde{r}$, если $r_1 = \widetilde{r_1} = 1, r_2 = \widetilde{r_2}, \dots, r_{i - 1} = \widetilde{r_{i-1}}$, причем $r_i \gt \widetilde{r_i}$


Для переходи от вектора $r$ к следующему (меньшему) вектору $\widetilde{r}$ необходимо найти самую старшую позицию, содержащую число больше 1 и уменьшаем ее на 1, предшествующие позиции оставляем без изменений.

### Алгоритм дробления индексов
Пусть задано $n \in \mathbb{N}$. Необходимо найти минимальную звездную цепочку, такую , что $a_m = n$.

Рассмотрим вектор индексов вида $$\left\{r_1,r_2,r_3,\dots,r_q\right\} \cup \left\{\rho_{q + 1},\rho_{q + 2},\rho_{q + 3},\dots,\rho_{m}\right\} $$ 
Назовем левую часть фиксированной, а правую - меняющейся.

$$a_{min} = a_{q+1} + m - q, \quad\text{при} \left\{r_1,r_2,r_3,\dots,r_q\right\} \cup \left\{1,1,1,\dots,1\right\} \\ 
a_{max} = a_{q+1} \cdot 2^{m - q}, \quad\text{при} \left\{r_1,r_2,r_3,\dots,r_q\right\} \cup \left\{q + 1,q + 2,q + 3,\dots,m\right\}$$

Описание алгоритма:
1. Запускаем внешний цикл по длинам цепочек: $\underline{l}(n) \le m \le \overline{l}(n)$.<br>
    Выбираем число $q \in \mathbb{N}$. Пусть $q = \frac{m}{2}$
2. Запускаем внутренний цекл перебора всех $\left\{r_1,r_2,r_3,\dots,r_q\right\}$ ($q!$ шагов). <br>
    На каждом шаге при фиксированной части вычисляем $a_{min}$ и $a_{max}$<br>
    1) Если $a_m = n$ - задача решена<br>
    2) Если $n \notin \left[a_{min},a_{max}\right]$, то переходим к следующему набору $\left\{r_1,r_2,r_3,\dots,r_q\right\}$<br>
    3) Если $n \in \left[a_{min},a_{max}\right]$, то организуем внутренний цикл перебора меняющейся части $\left\{\rho_{q + 1},\rho_{q + 2},\rho_{q + 3},\dots,\rho_{m}\right\}$. Таких наборов $\frac{m!}{q!}$ штук:<br>
    3.1) Если обнаруживается $a_m = n$ - задача решена.<br>
    3.2) Если в цикле таких векторов не оказалось, то переходим к следующей (по введенной упорядоченности) фиксированной части $\left\{r_1,r_2,r_3,\dots,r_q\right\}$
3. Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле.

### Гипотеза Шольца-Брауэра
$$l(2^{n} - 1) \le l(n) + n - 1$$
1. Гипотеза доказана для звездных цепочек:$$l^{*}(2^{n} - 1) \le l^{*}(n) + n - 1$$
2. Гипотеза справедлива для всех $n \lt 5784689$
3. Равенство выполняется для всех $1 \le n \le 64$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Выполнение работы

### 1) Алгоритм Брауэра
Реализуем функцию Brauer_alg, которая будет принимать число $n$, для которого необходимо найти приближенную аддитивную цепочку, и параметр $k$ и которая будет возвращать массив вспомогательных чисел и аддитивную цепочку.

In [1]:
def Brauer_alg(n, k):
    d = pow(2,k)
    q_arr = []
    r_arr = []
    q_tmp = n
    while q_tmp >= d:
        q_next = q_tmp // d
        q_arr.append(q_next)
        r_next = q_tmp % d
        r_arr.append(r_next)
        q_tmp = q_next
    result_arr = []
    helping_numbers = []
    for i in range(1,d,1):
        helping_numbers.append(i)
    tmp = q_arr[len(q_arr) - 1]
    for i in range(len(q_arr) - 1,-1, - 1):
        for j in range(k):
            tmp *= 2
            result_arr.append(tmp)
        tmp += r_arr[i]
        if r_arr[i] != 0:
            result_arr.append(tmp)
    while result_arr[0] < d:
        result_arr.pop(0)
    return (helping_numbers, result_arr)

Реализуем функции print_Brauer и print_Brauer_table, которые будут печатать результат построения аддитивной цепочки алгоритмом Брауэра, вторая функция необходима для эффективного построения таблицы.

In [207]:
def print_Brauer(result):
    print('Helping numbers: ',result[0])
    print('Additive chain: {},  Length:{}'.format(result[1], len(result[1])), '\n')
    
def print_Brauer_table(test_array):
    print('| Значение n | Значение k | Длина полученной цепочки | Аддитивная цепочка |')
    print('|------------|------------|--------------------------|---------------------|')
    for elem in test_array:
        for i in range(1,4,1):
            result = Brauer_alg(elem, i)
            chain = result[0]
            chain += result[1]
            print('|{}|{}|{}|{}|'.format(elem, i, len(chain), chain))
        

Протестируем функцию на различных $n$ и $k$:

In [208]:
test_array = [34563, 23462, 10231, 7567, 56786]
for elem in test_array:
    for i in range(2,5,1):
        print('n = {},  k = {}'.format(elem, i))
        print_Brauer(Brauer_alg(elem, i))

n = 34563,  k = 2
Helping numbers:  [1, 2, 3]
Additive chain: [4, 8, 16, 32, 33, 66, 132, 135, 270, 540, 1080, 2160, 4320, 8640, 17280, 34560, 34563],  Length:17 

n = 34563,  k = 3
Helping numbers:  [1, 2, 3, 4, 5, 6, 7]
Additive chain: [8, 16, 32, 64, 67, 134, 268, 536, 540, 1080, 2160, 4320, 8640, 17280, 34560, 34563],  Length:16 

n = 34563,  k = 4
Helping numbers:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
Additive chain: [16, 32, 64, 128, 135, 270, 540, 1080, 2160, 4320, 8640, 17280, 34560, 34563],  Length:14 

n = 23462,  k = 2
Helping numbers:  [1, 2, 3]
Additive chain: [4, 5, 10, 20, 22, 44, 88, 91, 182, 364, 366, 732, 1464, 1466, 2932, 5864, 5865, 11730, 23460, 23462],  Length:20 

n = 23462,  k = 3
Helping numbers:  [1, 2, 3, 4, 5, 6, 7]
Additive chain: [10, 20, 40, 45, 90, 180, 360, 366, 732, 1464, 2928, 2932, 5864, 11728, 23456, 23462],  Length:16 

n = 23462,  k = 4
Helping numbers:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
Additive chain: [20, 40, 80

Приведем таблицу аддитивных цепочек для различных $n$ и $k$:



| Значение n | Значение k | Длина полученной цепочки | Аддитивная цепочка |
|------------|------------|--------------------------|---------------------|
|34563|1|21|[1, 2, 4, 8, 16, 32, 33, 66, 67, 134, 135, 270, 540, 1080, 2160, 4320, 8640, 17280, 17281, 34562, 34563]|
|34563|2|20|[1, 2, 3, 4, 8, 16, 32, 33, 66, 132, 135, 270, 540, 1080, 2160, 4320, 8640, 17280, 34560, 34563]|
|34563|3|23|[1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 67, 134, 268, 536, 540, 1080, 2160, 4320, 8640, 17280, 34560, 34563]|
|23462|1|23|[1, 2, 4, 5, 10, 11, 22, 44, 45, 90, 91, 182, 183, 366, 732, 733, 1466, 2932, 5864, 5865, 11730, 11731, 23462]|
|23462|2|23|[1, 2, 3, 4, 5, 10, 20, 22, 44, 88, 91, 182, 364, 366, 732, 1464, 1466, 2932, 5864, 5865, 11730, 23460, 23462]|
|23462|3|23|[1, 2, 3, 4, 5, 6, 7, 10, 20, 40, 45, 90, 180, 360, 366, 732, 1464, 2928, 2932, 5864, 11728, 23456, 23462]|
|10231|1|24|[1, 2, 4, 8, 9, 18, 19, 38, 39, 78, 79, 158, 159, 318, 319, 638, 639, 1278, 2556, 2557, 5114, 5115, 10230, 10231]|
|10231|2|21|[1, 2, 3, 4, 8, 9, 18, 36, 39, 78, 156, 159, 318, 636, 639, 1278, 2556, 2557, 5114, 10228, 10231]|
|10231|3|22|[1, 2, 3, 4, 5, 6, 7, 8, 16, 19, 38, 76, 152, 159, 318, 636, 1272, 1278, 2556, 5112, 10224, 10231]|
|7567|1|21|[1, 2, 3, 6, 7, 14, 28, 29, 58, 59, 118, 236, 472, 944, 945, 1890, 1891, 3782, 3783, 7566, 7567]|
|7567|2|19|[1, 2, 3, 4, 7, 14, 28, 29, 58, 116, 118, 236, 472, 944, 1888, 1891, 3782, 7564, 7567]|
|7567|3|21|[1, 2, 3, 4, 5, 6, 7, 8, 14, 28, 56, 112, 118, 236, 472, 944, 945, 1890, 3780, 7560, 7567]|
|56786|1|25|[1, 2, 3, 6, 12, 13, 26, 27, 54, 55, 110, 220, 221, 442, 443, 886, 887, 1774, 3548, 3549, 7098, 14196, 28392, 28393, 56786]|
|56786|2|23|[1, 2, 3, 6, 12, 13, 26, 52, 55, 110, 220, 221, 442, 884, 887, 1774, 3548, 3549, 7098, 14196, 28392, 56784, 56786]|
|56786|3|25|[1, 2, 3, 4, 5, 6, 7, 8, 13, 26, 52, 104, 110, 220, 440, 880, 887, 1774, 3548, 7096, 7098, 14196, 28392, 56784, 56786]|


Приведем таблице минимальных аддитивных цепочек для тестовых чисел:


| Значение n | Длина минимальной аддитивной цепочки | Аддитивная цепочка | Длины цепочек полученных алгоритмом Брауэра |
|------------|--------------------------------------|--------------------|---------------------------------------------|
|34563|19|[1, 2, 3, 6, 12, 15, 30, 60, 120, 135, 270, 540, 1080, 2160, 4320, 8640, 17280, 34560, 34563]|[21, 20, 23]|
|23462|19|[1, 2, 4, 8, 16, 17, 33, 66, 83, 91, 182, 364, 728, 1456, 2912, 5824, 11648, 11731, 23462]|[23, 23, 23]|
|10231|18|[1, 2, 4, 5, 9, 14, 23, 37, 74, 148, 296, 319, 638, 1276, 2552, 5104, 10208, 10231]|[24, 21, 22]|
|7567|18|[1, 2, 3, 6, 7, 14, 28, 56, 59, 118, 236, 472, 944, 1888, 3776, 3783, 7566, 7567]|[21, 19, 21]|
|56786|21|[1, 2, 3, 6, 12, 13, 26, 52, 104, 208, 221, 442, 884, 1768, 3536, 3549, 7098, 14196, 28392, 28393, 56786]|[25, 23, 25]|


### Выводы по алгоритму Брауэра
Длины аддитивных цепочек, полученных с помощью алгоритма Брауэра не совпадают по длине с минимальными аддитивными цепочками, так как для работы алгоритма необходимо сначала вычислить "вспомогательные числа", которые не обязательно будут использованы. Алгоритм выдает приближенный ответ за малое время работы.


### 2) Алгоритм дробления вектора индексов
Для оценки длины аддитивных цепочек будем использовать следующие неравенства:
$$l(n) \le \lambda(n) + \upsilon(n) - 1\\
l(n) \ge \lambda(n) + \log_{2}(\upsilon(n))$$
Реализуем функцию a_constructor, которая по вектору индексов восстанавливает аддитивную цепочку. 

Реализуем функцию degree_vector_fractioning, которая будет принимать на вход число $n$ и искать для этого числа минимальную аддитивную цепочку: 

In [227]:
import math

def a_construction(chain):
    a_arr = [1]
    for elem in chain:
        a_arr.append(a_arr[len(a_arr) - 1] + a_arr[elem - 1])
    return (a_arr, a_arr[len(a_arr) - 1])
    
def degree_vector_fractioning(n):
    min_m = int(log(n,2)) + int(log(list(bin(n)).count('1'),2))
    max_m = int(log(n,2)) + list(bin(n)).count('1') - 1
    for m in range(min_m, max_m + 1, 1):
        q = m // 2
        r_arr = [i for i in range(1,q + 1,1)]
        for step in range(1, math.factorial(q) + 1, 1):
            a_q_1 = a_construction(r_arr)[1]
            a_min = a_q_1 + m - q
            a_max = a_q_1 * pow(2,m - q)
            if n >= a_min and n <= a_max:
                q_arr = [i for i in range(q + 1,m + 1,1)]
                for inter_step in range(1, math.factorial(m) // math.factorial(q) + 1, 1):
                    if(a_construction(r_arr + q_arr)[1] == n):
                        return(r_arr + q_arr)
                    for index in range(m - 1, q - 1, -1):
                        if q_arr[index - q] != 1:
                            q_arr[index - q] -= 1
                            break
                        else:
                            q_arr[index - q] = index + 1
                            
            for index in range(q - 1,  -1, -1):
                if r_arr[index] != 1:
                    r_arr[index] -= 1
                    break
                else:
                    r_arr[index] = index + 1


Рассмотрим работу функции на тестовом массиве чисел, числа выбраны специальным образом, чтобы время поиска $n$ не равнялось половине часа:

In [228]:
from datetime import datetime
safe_test_array = [1152, 1600, 1344, 1024, 1288]
for elem in safe_test_array:
    t = datetime.now()
    index_vector = degree_vector_fractioning(elem)
    t = datetime.now() - t
    print('n = {})\n\tIndex vector: {}\n\tAdditive chain: {}  Length: {}\n\tTime: {}\n\n'.format(elem,index_vector, a_construction(index_vector)[0], len(a_construction(index_vector)[0]), t))

n = 1152)
	Index vector: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 8]
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152]  Length: 12
	Time: 0:00:00.000804


n = 1600)
	Index vector: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 7]
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1600]  Length: 13
	Time: 0:00:03.971590


n = 1344)
	Index vector: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 7]
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1344]  Length: 13
	Time: 0:00:18.795920


n = 1024)
	Index vector: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]  Length: 11
	Time: 0:00:00.000184


n = 1288)
	Index vector: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 4]
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1288]  Length: 13
	Time: 0:00:18.938323




Сравним результаты работы полного перебора с алгоритмом Брауэра:

In [221]:
for elem in safe_test_array:
        for i in range(1,4,1):
            result = Brauer_alg(elem, i)
            chain = result[0]
            chain += result[1]
            print('n = {}, k = {})\n\tAdditive chain: {}  Length: {}\n'.format(elem, i, chain, len(chain)))

n = 1152, k = 1)
	Additive chain: [1, 2, 4, 8, 9, 18, 36, 72, 144, 288, 576, 1152]  Length: 12

n = 1152, k = 2)
	Additive chain: [1, 2, 3, 4, 8, 16, 18, 36, 72, 144, 288, 576, 1152]  Length: 13

n = 1152, k = 3)
	Additive chain: [1, 2, 3, 4, 5, 6, 7, 8, 16, 18, 36, 72, 144, 288, 576, 1152]  Length: 16

n = 1600, k = 1)
	Additive chain: [1, 2, 3, 6, 12, 24, 25, 50, 100, 200, 400, 800, 1600]  Length: 13

n = 1600, k = 2)
	Additive chain: [1, 2, 3, 4, 6, 12, 24, 25, 50, 100, 200, 400, 800, 1600]  Length: 14

n = 1600, k = 3)
	Additive chain: [1, 2, 3, 4, 5, 6, 7, 12, 24, 25, 50, 100, 200, 400, 800, 1600]  Length: 16

n = 1344, k = 1)
	Additive chain: [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344]  Length: 13

n = 1344, k = 2)
	Additive chain: [1, 2, 3, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344]  Length: 14

n = 1344, k = 3)
	Additive chain: [1, 2, 3, 4, 5, 6, 7, 8, 16, 21, 42, 84, 168, 336, 672, 1344]  Length: 16

n = 1024, k = 1)
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 128,

|Значение n|Вектор индексов|Аддитивная цепочка|Длина минимальной аддитивной цепочки|Время работы|Длины в Брауэре|
|-|-|-|-|-|-|
|1152|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 8]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152]|12|0:00:00.000866|[12, 13, 16]|
|1600|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 7]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1600]|13|0:00:04.368692|[13, 14, 16]|
|1344|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 7]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1344]|13|0:00:20.901092|[13, 14, 16]|
|1024|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]|11|0:00:00.000165|[11, 12, 15]|
|1288|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 9, 4]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1288]|13|0:00:21.122585|[13, 14, 17]|

### Выводы по алгоритму дробления вектора индексов
Алгоритм позволяет получить минимальную аддитивную цепочку за счет того, что осуществляется полный перебор. Минусом этого алгоритма является невероятно большое время работы, которое является таковым по этой же причине.

### 3) Гипотеза Шольца-Брауэра
Проверим гипотезу Шольца-Брауэра на числах $1 \le n \le 12$, которая гласит, что для этих чисел должно выполняться равенство: 
$$l(2^{n} - 1) = l(n) + n - 1$$

In [ ]:
Scholz_array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
for elem in Scholz_array:
    index_vector = degree_vector_fractioning(pow(2,elem) - 1)
    additive_chain = a_construction(index_vector)[0]
    print('n = {})\n\tAdditive chain: {}  \n\tLength of (2^n - 1) and (n): {}, {}\n'.format(elem,additive_chain,len(additive_chain), len(a_construction(degree_vector_fractioning(elem))[0])))

n = 1)
	Additive chain: [1]  
	Length of (2^n - 1) and (n): 1, 1

n = 2)
	Additive chain: [1, 2, 3]  
	Length of (2^n - 1) and (n): 3, 2

n = 3)
	Additive chain: [1, 2, 4, 6, 7]  
	Length of (2^n - 1) and (n): 5, 3

n = 4)
	Additive chain: [1, 2, 4, 5, 10, 15]  
	Length of (2^n - 1) and (n): 6, 3

n = 5)
	Additive chain: [1, 2, 4, 8, 10, 20, 30, 31]  
	Length of (2^n - 1) and (n): 8, 4

n = 6)
	Additive chain: [1, 2, 4, 8, 16, 20, 21, 42, 63]  
	Length of (2^n - 1) and (n): 9, 4

n = 7)
	Additive chain: [1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127]  
	Length of (2^n - 1) and (n): 11, 5

n = 8)
	Additive chain: [1, 2, 4, 8, 16, 17, 34, 68, 85, 170, 255]  
	Length of (2^n - 1) and (n): 11, 4

n = 9)
	Additive chain: [1, 2, 4, 8, 16, 32, 64, 72, 73, 146, 292, 438, 511]  
	Length of (2^n - 1) and (n): 13, 5



|Значение (n^2 - 1)|Аддитивная цепочка|Длина минимальной аддитивной цепочки|Значение n|Длина аддитивной цепочки для n|Теоритическая длина по гипотезе|
|----------------|------------------|------------------------------------|----------|------------------------------|----------------------------------|
|1|[1]|1|1|1|1 + 1 - 1 = 1|
|3|[1, 2, 3]|3|2|2|2 + 2 -1 =3|
|7|[1, 2, 4, 6, 7]|5|3|3|3 + 3 - 1 = 5|
|15|[1, 2, 4, 5, 10, 15]|6|4|3|4 + 3 - 1 = 6|
|31|[1, 2, 4, 8, 10, 20, 30, 31]|8|5|4|5 + 4 - 1 = 8|
|63|[1, 2, 4, 8, 16, 20, 21, 42, 63]|9|6|4|6 + 4 - 1 = 9|
|127|[1, 2, 4, 8, 16, 32, 40, 42, 84, 126, 127]|11|7|5|7 + 5 - 1 = 11|
|255|[1, 2, 4, 8, 16, 17, 34, 68, 85, 170, 255]|11|8|4|8 + 4 - 1 = 11|
|511|[1, 2, 4, 8, 16, 32, 64, 72, 73, 146, 292, 438, 511]|13|9|5|9 + 5 - 1 = 13|
|1023|[1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 1020, 1023]|14|10|5|10 + 5 - 1 = 14|
|2047|[1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 1020, 1023, 2046, 2047]|16|11|6|11 + 6 - 1 = 16|
|4095|[1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 1020, 2040, 4080, 4095]|16|12|5|12 + 5 - 1 = 16|


### Выводы по Гипотезе Шольца-Брауэра
Для $1 \le n \le 12$ была проверена и доказана подлинность гипотезы.

## Выводы

Реализованы алгоритм Брауэра и алгоритм дробления вектора индексов, сравнены между собой.Применены навыкы использования системы компьютерной математики SageMath. Так как нет эффективного способа вычисления минимальной аддитивной цепочки, то приходится выбирать между алгоритмом приближенным, работающим за малое время, либо алгоритмом точным, но работающим за невероятно большое время.